### PR TITLE
Pset4

### PR DESCRIPTION
**Pull request by hljames**

### CODE REVIEW

File: checks/cs50/2017/x/recover/checks.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    @check("compiles")
+    def test_noimage(self):
+        """handles lack of forensic image"""
+        self.spawn("./recover").exit(1)
+
+    @check("compiles")
+    def first_image(self):
+        """recovers 000.jpg correctly"""
+        self.include("card.raw")
+        subprocess.call("bash", shell=True)
```
Comment (dmalan) [inline checks/cs50/2017/x/recover/checks.py]: What's this for?

File: checks/cs50/2017/x/recover/checks.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        self.spawn("./recover card.raw").exit(0)
+        hash000 = 'c3ceda9dde5cf2212e96d6f61e2e849ac60e2c3cbdf0184f6e0c3b6ea5e69d31'
```
Comment (dmalan) [inline checks/cs50/2017/x/recover/checks.py]: Double quotes for consistency.

File: checks/cs50/2017/x/recover/checks.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        if self.hash("000.jpg") != hash000:
```
Comment (dmalan) [inline checks/cs50/2017/x/recover/checks.py]: Not sure we gain much by having the variable.

File: checks/cs50/2017/x/recover/checks.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            raise Error("recovered image does not match")
+
+    @check("compiles")
+    def middle_images(self):
+        """recovers middle images correctly"""
+        self.include("card.raw")
```
Comment (dmalan) [inline checks/cs50/2017/x/recover/checks.py]: @crossroads1112 should the semantics here perhaps be `self.copy` or `self.import` instead? Not sure, just had to think twice about what `include` meant here.

File: checks/cs50/2017/x/recover/checks.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        self.spawn("./recover card.raw").exit(0)
+        middlehashes = ['8c0bd07a6c232670fd55b50ac42611c349920a119ef575eb28f770b9c89753e4',
```
Comment (dmalan) [inline checks/cs50/2017/x/recover/checks.py]: Double quotes.

File: checks/cs50/2017/x/recover/checks.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            'f22fd0a4e18f076c043bd8306fe771f2fa843f764e5c9c399a8f61d4921c1743',
+            '5933fbd4d2ce7194ed67241c1f65fc86b5c882884de63644ceef4cd6ec97dd4c',
+            'cbf1f76713e3a0c6b2779f9a3489291783ee9e8f4ec7469d5be362e892cbfbdd',
+            '24b2b88477534ee29f7370d0742c172bef083a8fa78505cbc4bf7fd7df3353ca',
+            '2f1900342cfa54c0b5be9bc65d3dedbdcc2fe48c105b75700de1087ba93f581a',
+            'e13ddf43b9524f0cfdf23b8ac89b1bcdc21bb9030d5a80637a511f84c2c807e6',
+            'ff87370119a48b8f78b8ff96476bfb1f0a317339922eb6d526859674f1349bd7']
+        c = 0
+        while (c < 14):
+            h = self.hash("0" + str(c + 1).zfill(2) + ".jpg")
```
Comment (dmalan) [inline checks/cs50/2017/x/recover/checks.py]: This block should definitely be commented.

File: checks/cs50/2017/x/recover/checks.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            if (h != middlehashes[c]):
+                raise Error("recovered image does not match")
+            c += 1
+
+    @check("compiles")
+    def last_image(self):
+        """recovers 015.jpg correctly"""
+        self.include("card.raw")
+        self.spawn("./recover card.raw").exit(0)
+        hash015 = '49442513773931634fc572de401bb44d6d91069b23113e86a16336f07d69021d'
```
Comment (dmalan) [inline checks/cs50/2017/x/recover/checks.py]: Quotes.

File: checks/cs50/2017/x/recover/checks.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        if self.hash("015.jpg") != hash015:
```
Comment (dmalan) [inline checks/cs50/2017/x/recover/checks.py]: Var unneeded.

File: checks/cs50/2017/x/resize/less/checks.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+sys.path.append(os.getcwd())
+from check50 import *
+
+class Resize_Less(TestCase):
+
+    @check()
+    def resize_bmp_exist(self):
+        """resize.c and bmp.h exist."""
+        super().exists("resize.c")
```
Comment (dmalan) [inline checks/cs50/2017/x/resize/less/checks.py]: @crossroads1112 can we change `exists` to take a `list` of files too?

File: checks/cs50/2017/x/resize/less/checks.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        child = self.spawn("./resize 1 1.bmp outfile.bmp")
+        self.spawn("diff 1.bmp outfile.bmp").stdout("", "").exit(0)
+
+    @check("compiles")
+    def scale_by_2(self):
+        """resizes 1x1-pixel BMP to 2x2 correctly when n is 2"""
+        self.include("1.bmp")
+        self.include("2.bmp")
+        child = self.spawn("./resize 2 1.bmp outfile.bmp")
+        self.spawn("diff 2.bmp outfile.bmp").stdout("", "").exit(0)
```
Comment (dmalan) [inline checks/cs50/2017/x/resize/less/checks.py]: Why expecting `""` here and elsewhere?

File: checks/fifteen.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        import pexpect
+        child = pexpect.spawn("./fifteen 3")
+        child.expect(".+")
+        import time
+        print(os.getcwd(), file=sys.stderr)
+        self.include("outputs/fifteen/init-3x3.txt")
+        # child  = subprocess.Popen("./fifteen 3", shell=True)
+        # time.sleep(3)
+        # child.kill()
+        subprocess.Popen("bash", shell=True).wait()
```
Comment (dmalan) [inline checks/fifteen.py]: What's this?

File: checks/cs50/2017/x/resize/less/checks.py (hunk 9 above)
Comment (brianyu28) [inline checks/cs50/2017/x/resize/less/checks.py]: Already done per https://github.com/cs50/check50/commit/e2da3f2c0dcc4892f317d3c3f66749ba9696a188!

Comment (dmalan) [inline checks/cs50/2017/x/resize/less/checks.py]: Thanks, @crossroads1112! :)

File: checks/cs50/2017/x/recover/checks.py (hunk 4 above)
Comment (crossroads1112) [inline checks/cs50/2017/x/recover/checks.py]: Copy feels semantically like it takes two arguments. `self.copy("card.raw")` seems to raise the question "Copy card.raw to/from where?"

Import doesn't feel quite right either. At least in the pythonic sense, `import` feels like it implies that I'm going execute `card.raw` somehow to facilitate the check. 

I think `include` is the right semantic. It's sort of saying that when gathering all the stuff we need for this check, we must `include card.raw`. It also evokes `#include` which is a sort of copy/paste operation.

File: checks/cs50/2017/x/resize/less/checks.py (hunk 10 above)
Comment (crossroads1112) [inline checks/cs50/2017/x/resize/less/checks.py]: I believe it means we are expecting no output i.e. diff outputs nothing. The exit code is probably a better indicator though. `diff` returns a 0 when the files are identical, 1 when they're different or a 2 on error (file not found).

File: checks/fifteen.py (hunk 11 above)
Comment (crossroads1112) [inline checks/fifteen.py]: IIRC, this is an artifact of debugging the check. Should be removed.

File: checks/cs50/2017/x/recover/checks.py (hunk 4 above)
Comment (dmalan) [inline checks/cs50/2017/x/recover/checks.py]: 👍

File: checks/cs50/2017/x/resize/less/checks.py (hunk 10 above)
Comment (dmalan) [inline checks/cs50/2017/x/resize/less/checks.py]: Didn't we introduce `.stdout(self.EOF)` or such for this?

File: checks/cs50/2017/x/recover/checks.py (hunk 1 above)
Comment (crossroads1112) [inline checks/cs50/2017/x/recover/checks.py]: Artifact of debugging I think, will remove

**Pull request merged**